### PR TITLE
Skeleton preview

### PR DESCRIPTION
**Pull request by ImUrX**
Based on #787

### CODE REVIEW

**Comment by ImUrX** [comment]
Ready for review

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx [hunk 1]
```diff
@@ -1,122 +1,132 @@
...
-  leftLowerLeg.add(leftFoot);
+const yParts = [
+  BodyPart.NECK,
+  BodyPart.UPPER_CHEST,
+  BodyPart.CHEST,
+  BodyPart.WAIST,
+  BodyPart.HIP,
+  BodyPart.LEFT_UPPER_LEG,
+  BodyPart.LEFT_LOWER_LEG,
+];
```
Comment (ButterscotchV) [inline gui/src/components/widgets/SkeletonVisualizerWidget.tsx:56]: Might be worth a comment explaining why right leg is missing. I'm assuming this is just to get the total height?

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx [hunk 2]
```diff
@@ -1,122 +1,132 @@
...
+    );
+  }, [bones]);
 
-      if (!bone.parent) {
-        bone.applyQuaternion(QuaternionFromQuatT(solarBone.rotationG));
-        continue;
-      }
+  const heightOffset = useMemo(() => {
+    const hmd = bones.get(BodyPart.HEAD);
+    if (hmd?.headPositionG?.y === undefined || hmd.headPositionG?.y > 0) {
```
Comment (ButterscotchV) [inline gui/src/components/widgets/SkeletonVisualizerWidget.tsx:80]: This looks really weird, why is it returning 0 if it's undefined? What's the height offset for, when the head position is unavailable?

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx (hunk 1 above)
Comment (ImUrX) [reply gui/src/components/widgets/SkeletonVisualizerWidget.tsx:56]: yep

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx (hunk 2 above)
Comment (ImUrX) [reply gui/src/components/widgets/SkeletonVisualizerWidget.tsx:80]: yep

**Comment by ButterscotchV** [comment]
Some quick notes from testing, it's hard to differentiate the chest and upper legs from the background, shoulders are a bit questionable. Should probably have a defined background colour (if there isn't already), maybe with a light and dark mode option.
image

Otherwise the FPS feels really crunchy lol. Small complaint but maybe just a smidgen of interpolation or smth would make it look better, idk how good of an idea it is, would appreciate feedback on it.

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx [hunk 3]
```diff
@@ -1,122 +1,144 @@
...
+declare module '@react-three/fiber' {
+  interface ThreeElements {
+    basedSkeletonHelper: Object3DNode<
+      BasedSkeletonHelper,
+      typeof BasedSkeletonHelper
+    >;
+  }
+}
+
+const groundColor = '#4444aa';
```
Comment (loucass003) [inline gui/src/components/widgets/SkeletonVisualizerWidget.tsx:25]: Const outside components funtions should be uppercase

File: gui/src/components/widgets/SkeletonVisualizerWidget.tsx [hunk 4]
```diff
@@ -1,122 +1,144 @@
...
+      resolution={res}
+      args={[object]}
+    />
   );
+}
 
-  return new Skeleton(bones);
-})();
+// Just need to know the length of the total body, so don't need right legs
+const yParts = [
```
Comment (loucass003) [inline gui/src/components/widgets/SkeletonVisualizerWidget.tsx:62]: Const outside components funtions should be uppercase

**Review by loucass003 — APPROVED** [review]
Code looks good, not tested tho

**Review by ButterscotchV — APPROVED** [review]
Asside from code specifics that I can't comment on, this looks good. I skimmed the code and it looks fine, and tested lightly and it seems to work fine. I haven't tested thoroughly with the line thickness fix, but I imagine it's fine.

**Review by Louka3000 — CHANGES_REQUESTED** [review]
Foot shift/foot length is still broken:
slimevr_qg3vNQEv0T

The renderer seems easily confused about whether the skeleton is in front of the floor grid or not

The camera should be "looking at" the front of the head ideally.

**Comment by ImUrX** [comment]
I wont do the camera change, that's gonna be agonizing to see.
The rest, I will check if I can fix

**Comment by Louka3000** [comment]
> I wont do the camera change, that's gonna be agonizing to see.

Why? To be clear, I mean to make it "look at the head" initially. Not constantly looking at it.

**Comment by ImUrX** [comment]
ow, then sorry I misunderstood www

File: gui/src/utils/skeletonHelper.ts [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    // console.log(this.quaternion);
+    // console.log(
+    //   parentBone === undefined
+    //     ? new Vector3(0, 0, 0)
+    //     : Vector3FromVec3fT(parentBone.headPositionG),
+    //   Vector3FromVec3fT(this.boneT.headPositionG)
+    // );
+    this.position.set(0, -this.boneT.boneLength, 0);
+    this.position.applyQuaternion(quat);
+    // console.log(this.position);
```
[the remark this answers is not in the document]
Comment (ImUrX) [reply gui/src/utils/skeletonHelper.ts:183]: these are useful for debugging, I will keep them for now

**Review by Louka3000 — APPROVED** [review]

**Review by Eirenliel — APPROVED** [review]

**Review by loucass003 — APPROVED** [review]

**Pull request merged**